sklearn/cluster/tests/test_k_means.py

```diff
@@ -160,21 +161,18 @@ def test_kmeans_elkan_results(distribution, array_constr, tol):
     assert km_elkan.inertia_ == pytest.approx(km_full.inertia_, rel=1e-6)
 
 
-@pytest.mark.parametrize("algorithm", ["full", "elkan"])
-def test_kmeans_convergence(algorithm):
+def test_kmeans_convergence():
     # Check that KMeans stops when convergence is reached when tol=0. (#16075)
-    # We can only ensure that if the number of threads is not to large,
-    # otherwise the roundings errors coming from the unpredictability of
-    # the order in which chunks are processed make the convergence criterion
-    # to never be exactly 0.
     rnd = np.random.RandomState(0)
     X = rnd.normal(size=(5000, 10))
+    max_iter = 300
 
-    with threadpool_limits(limits=1, user_api="openmp"):
-        km = KMeans(algorithm=algorithm, n_clusters=5, random_state=0,
-                    n_init=1, tol=0, max_iter=300).fit(X)
+    km_full = KMeans(algorithm="full", n_clusters=5, random_state=0,
+                     n_init=1, tol=0, max_iter=max_iter).fit(X)
+    km_elkan = KMeans(algorithm="elkan", n_clusters=5, random_state=0,
+                      n_init=1, tol=0, max_iter=max_iter).fit(X)
 
-    assert km.n_iter_ < 300
+    assert km_full.n_iter_ == km_elkan.n_iter_ < max_iter
```

Comment:
Will this assertion be robust to rounding errors that can affect Lloyd and Elkan slightly differently?

Reply:
That is a badly broken line anyway that is likely equivalent to
```
    assert int(km_full.n_iter_ == km_elkan.n_iter_) < max_iter
```
isn't it? And because int(boolean) is either 0 or 1, this will likely always be True (unless you use max_iter=1 in some test).

As for numerical issues: in theory it can happen that because of rounding issues Lloyd and Elkan in a rare case may produce different results. But I would rather test for them to do the same number of iterations and then study the situation if they ever *not* agree; if ever. That would be an interesting case to see.

Reply:
It's not broken. In python you can chain comparison operators. ``a  == b < c`` is equivalent to ``a == b and b < c``.

It's not robust to rounding errors indeed. While I agree that it's interesting to check when such a situation happens, it's painful to have that in the test suite, because it's hard to debug when it suddently fails on the ci because some hardware change or some other reason idk. It happened with another test and we just ended up xfailing the test...

Reply:
also there's already a test which checks that elkan and lloyd have the same n_iter_ in a situation where rounding errors can't interfere (test_kmeans_elkan_results iirc)